src/System.Linq/tests/SingleTests.cs

```diff
@@ -0,0 +1,50 @@
+﻿using System.Linq.Tests.Helpers;
+using Xunit;
+
+namespace System.Linq.Tests
+{
+    public class SingleTests
+    {
+        [Fact]
+        public void FailOnEmpty()
+        {
+            Assert.Throws<InvalidOperationException>(() => Enumerable.Empty<int>().Single());
+        }
+
+        [Fact]
+        public void DefaultOnEmpty()
+        {
+            Assert.Equal(0, Enumerable.Empty<int>().SingleOrDefault());
+        }
+
+        // Warning: At time of writing these two tests, they will incorrectly run for about five millennia.
+        // They'll then probably throw OverflowException rather than the correct InvalidOperationException,
+        // but I haven't had time to confirm. Recommend to skip if using without fix applied in commit 3f17916
+
+        [Fact]
+        public void FailOnManyMatches()
+        {
+            Assert.Throws<InvalidOperationException>(() => Enumerable.Range(0, int.MaxValue).Single(i => i > 0));
```

Comment:
I would prefer to see this done differently.  I'd have a test like:
```C#
[Fact]
public void DelegateInvokedForAllMatches()
{
    int count = 0;
    int iters = 10;
    Assert.Equal(1, Enumerable.Range(1, iters).Single(i => { count++; return true; }));
    Assert.Equal(iters, count);
}
```
and I'd get that merged with the rest of the tests.  That helps verify the current behavior, which someone could be depending on (though that'd be unfortunate).  Then separately there can be a discussion about changing that behavior, and if/when a fix for it goes in, the test can be updated to reflect that count is now 2 instead of 10. Etc.